Item 6835: Data Class designer updates for use in Sample Manager

cnathe opened the pull request.

cnathe committed:
Item 6835: Data Class Designer (aka "Source" in Sample Manager)

cnathe committed:
Item 6835: remove redundant check for data class sample set property exists on create/update

cnathe committed:
Item 6835: expose DataClassCategoryType as table in exp user schema for use with selectRows to populate dropdown

cnathe committed:
Item 6835: update to latest @labkey/components alpha package

cnathe committed:
Item 6835: fixes for DomainDesigner.tsx related to handling error messages (we only expect one so no need for a List<IBannerMessage>)

cnathe committed:
Item 6835: add type for DataClassDesigner.tsx fetchDataClass callback param

cnathe committed:
Merge branch 'develop' into fb_smSourceCRUD

cnathe committed:
Merge branch 'develop' into fb_smSourceCRUD

cnathe committed:
Item 6835: Update to latest @labkey/components alpha package version

cnathe committed:
Item 6835: add dataClassDesigner generated views to gitignore file

cnathe committed:
Merge branch 'develop' into fb_smSourceCRUD

cnathe committed:
Update to latest @labkey/components alpha package version

cnathe committed:
Merge branch 'develop' into fb_smSourceCRUD

cnathe committed:
Item 6835: use DomainUtil.validateProperties for reserved fields check on data class creation

cnathe committed:
Item 6835: minor update to entryPoints.js related to permissions check for dataClassDesigner view

cnathe committed:
Issue 39917: Manually adding a field to an existing Source Type results in Created date and Created By for the Source to be cleared

cnathe committed:
Merge branch 'develop' into fb_smSourceCRUD

cnathe committed:
Update to latest @labkey/components alpha package version

cnathe committed:
Merge branch 'develop' into fb_smSourceCRUD

# Conflicts:
#	assay/package-lock.json
#	assay/package.json
#	core/package-lock.json
#	core/package.json
#	experiment/package-lock.json
#	experiment/package.json
#	list/package-lock.json
#	list/package.json

cnathe committed:
Update to latest @labkey/components alpha package version

Sigmonia commented on api/src/org/labkey/api/exp/api/DataClassDomainKindProperties.java:
```diff
@@ -0,0 +1,112 @@
+package org.labkey.api.exp.api;
+
+public class DataClassDomainKindProperties
```
Better name for these: DataClassPropertiesModel ?

cnathe replied:
@labkey-ians  Yes that is probably better. I was just following the name from ListDomainKindProperties. Did you use a different name for your sample type conversion work?

Sigmonia replied:
Nope did the same as you, followed List. But i'm game to make the change in my branch too.

cnathe replied:
We can refactor them together after both these branches are merged.

Sigmonia commented on api/src/org/labkey/api/exp/api/DataClassDomainKindProperties.java:
```diff
@@ -0,0 +1,112 @@
+
+    public String getName()
+    {
+        return name;
```
Should these String getters be StringUtils.trimToNull(name)?

cnathe replied:
I don't think this would be the right place for that, these are just the bean getters/settings. I'd prefer if the trimToNull is done when pulling the values out of the options object before save. Thoughts?

Sigmonia replied:
I tend to agree, my only concern is not calling it when necessary. This gives a chokepoint to reduce the chance that trim wouldn't be applied.

cnathe replied:
Good point. I'll give it a go.

Sigmonia commented on experiment/src/client/DataClassDesigner/DataClassDesigner.tsx:
```diff
@@ -0,0 +1,112 @@
+            this.setState(() => ({isLoading: false}));
+        }
+
+        window.addEventListener("beforeunload", this.handleWindowBeforeUnload);
```
Would be nice to pull this, the remove, and the supporting methods into a HOC

cnathe replied:
Good idea. Done.

cnathe committed:
Fixes for test failures related to domain reserved fields checks

Sigmonia commented on experiment/src/org/labkey/experiment/api/DataClassDomainKind.java:
```diff
@@ -290,4 +275,12 @@ public String getObjectUriColumnName()
+
+    @Nullable
+    @Override
+    public DataClassDomainKindProperties getDomainKindProperties(GWTDomain domain, Container container, User user)
```
```suggestion
    public DataClassDomainKindProperties getDomainKindProperties(@NotNull GWTDomain domain, Container container, User user)
```

cnathe replied:
Done.

Sigmonia commented on experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java:
```diff
@@ -6296,37 +6296,21 @@ public ExpDataClassImpl createDataClass(
-            if (ss == null)
-                throw new IllegalArgumentException("SampleSet '" + sampleSetId + "' not found");
+        DataClassDomainKindProperties options = new DataClassDomainKindProperties();
+        options.setDescription(description);
```
We should be consistent with this. For ST I used the description field that is part of the GWTDomain.

cnathe replied:
But the top level DataClass object is the one that has the description field. I saw that domain description field usage in your ST branch. For data classes, the description for a create and for update go into the top level DataClass object (from what I can tell). As your next comment mentions, I saw the usage of domain.getDescription in the previous DataClassDomainKind.createDomain code, but I can't figure out how that would have ever been populated on that domain object in the first place.

Sigmonia commented on experiment/src/org/labkey/experiment/api/DataClassDomainKind.java:
```diff
@@ -199,56 +204,36 @@ public boolean canCreateDefinition(User user, Container container)
-        if (name == null)
-            throw new IllegalArgumentException("DataClass name required");
-
-        String description = domain.getDescription();
```
We should be consistent with this. For ST I used the description field that is part of the GWTDomain. Primarily because this was already in place.

cnathe replied:
See comment above.

Sigmonia commented on experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java:
```diff
@@ -6355,34 +6333,36 @@ else if (domain.getPropertyByName(propertyName) != null) // issue 25275
             domain.save(u);
             impl.save(u);
+
+            //TODO do DataClasses actually support default values? The DataClassDomainKind does not override showDefaultValueSettings to return true so it isn't shown in the UI.
```
Still relevant?

cnathe replied:
Yes. I don't know the answer and need to follow-up with Kevin (will likely do that during the LKS side of this designer work).

Sigmonia commented on experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java:
```diff
@@ -6392,6 +6372,77 @@ else if (domain.getPropertyByName(propertyName) != null) // issue 25275
+        if (options != null)
+        {
+            validateDataClassOptions(c, u, options);
+            dataClass.setDescription(options.getDescription());
```
Again should make this consistent across our various domain types

cnathe replied:
See comments above.

Sigmonia commented on experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java:
```diff
@@ -6392,6 +6372,77 @@ else if (domain.getPropertyByName(propertyName) != null) // issue 25275
+
+        TableInfo dataClassTable = ExperimentService.get().getTinfoDataClass();
+        int nameExpMax = dataClassTable.getColumn("NameExpression").getScale();
+        if (options.getNameExpression() != null && options.getNameExpression().length() > nameExpMax)
```
Do these string values need to be cleaned up? StringUtils.trimToNull?

cnathe replied:
Done. From the comment above, adding the trimToNull inside of the getters for DataClassDomainKindProperties did the trick.

Sigmonia commented on experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java:
```diff
@@ -6392,6 +6372,77 @@ else if (domain.getPropertyByName(propertyName) != null) // issue 25275
+
+        // Validate category length
+        int categoryMax = dataClassTable.getColumn("Category").getScale();
+        if (options.getCategory() != null && options.getCategory().length() > categoryMax)
```
same trimToNull?

cnathe replied:
Done. From the comment above, adding the trimToNull inside of the getters for DataClassDomainKindProperties did the trick.

cnathe committed:
Merge branch 'develop' into fb_smSourceCRUD

# Conflicts:
#	assay/package-lock.json
#	assay/package.json
#	core/package-lock.json
#	core/package.json
#	experiment/package-lock.json
#	experiment/package.json
#	list/package-lock.json
#	list/package.json

cnathe committed:
Merge branch 'develop' into fb_smSourceCRUD

# Conflicts:
#	core/package-lock.json
#	core/package.json

cnathe committed:
Update to latest @labkey/components alpha package version

cnathe committed:
Item 6835: Fix for Biologics case for creating DataClasses from a template
- where the template has the sampleSet name instead of RowId so the DataClassDomainKindProperties object mapping was getting mad

labkey-danield committed:
Merge branch 'develop' into fb_smSourceCRUD

labkey-danield committed:
Merge branch 'develop' into fb_smSourceCRUD

cnathe committed:
Update to latest @labkey/components alpha package version

cnathe committed:
Update to latest @labkey/components alpha package version

cnathe committed:
Update to latest @labkey/components alpha package version

cnathe committed:
DataClassDesigner.tsx updates to use BeforeUnload HOC

cnathe committed:
DataClassDomainKindProperties updates to use StringUtils.trimToNull

cnathe committed:
Merge branch 'develop' into fb_smSourceCRUD

# Conflicts:
#	assay/package-lock.json
#	assay/package.json
#	core/package-lock.json
#	core/package.json
#	experiment/package-lock.json
#	experiment/package.json
#	experiment/src/client/entryPoints.js
#	list/package-lock.json
#	list/package.json

cnathe committed:
Update to latest @labkey/components package version

cnathe committed:
Update to @labkey/components version 0.39.0
